### PR TITLE
New function to get compressed files

### PR DESCRIPTION
**Pull request by CGodiksen**
This PR includes a new `get_compressed_files` function in the storage engine that makes it possible to return compressed files that correspond to a list of 64-bit keys. The function also takes a start time and an end time to support pruning the files as early as possible. To support pruning both based on start time and end time, the compressed file names have been changed to include the entire time range the file covers and not just the start time. This means compressed files now have the file name format `start_timestamp-end_timestamp.parquet`. 

The PR also contains testing for the new function in the storage engine tests and tests for the utility functions created in the compressed data manager. 

It should be noted that this branch was created on top of the [storage-engine-ingestion](https://github.com/ModelarData/ModelarDB-RS/pull/30) branch which means that only the latest 12 commits in this PR are new.

### CODE REVIEW

File: server/src/storage/time_series.rs [hunk 1]
```diff
@@ -74,13 +75,20 @@ impl CompressedTimeSeries {
...
+        // name to better support pruning data that is too new or too old when executing a specific query.
         let start_times: &TimestampArray = batch.column(2).as_any().downcast_ref().unwrap();
-        let file_name = format!("{}.parquet", start_times.value(0));
-        let file_path = complete_folder_path.join(file_name);
+        let end_times: &TimestampArray = batch.column(3).as_any().downcast_ref().unwrap();
 
-        StorageEngine::write_batch_to_apache_parquet_file(batch, file_path.as_path());
+        let file_name = format!(
+            "{}-{}.parquet",
+            start_times.value(0),
```
Comment (skejserjensen) [inline server/src/storage/time_series.rs:85]: Maybe add a doc comment to `append_segment()` that states that `segment: RecordBatch` is assumed to be sorted by time.

File: server/src/storage/mod.rs [hunk 2]
```diff
@@ -189,9 +242,141 @@ impl StorageEngine {
 mod tests {
     use super::*;
     use std::path::PathBuf;
+    use std::thread;
+    use std::time::Duration;
 
+    use datafusion::arrow::array::ArrayAccessor;
```
Comment (skejserjensen) [inline server/src/storage/mod.rs:248]: Unclear why `ArrayAccessor` must be imported.

**Review by skejserjensen — CHANGES_REQUESTED** [review]

File: server/src/storage/compressed_data_manager.rs [hunk 3]
```diff
@@ -81,34 +85,104 @@ impl CompressedDataManager {
 
         // If the reserved memory limit is exceeded, save compressed data to disk.
         if self.compressed_remaining_memory_in_bytes < 0 {
-            self.save_compressed_data();
+            self.free_compressed_data_memory();
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/compressed_data_manager.rs:88]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 4]
```diff
@@ -81,34 +85,104 @@ impl CompressedDataManager {
...
         }
     }
 
+    /// Return the file path to each on-disk compressed file that corresponds to `key`. If some
+    /// compressed data that corresponds to `key` is still in memory, save the data to disk first.
+    /// If `key` does not correspond to any data, [`DataRetrievalError`](ModelarDBError::DataRetrievalError) is returned.
+    pub fn get_saved_compressed_files(
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/compressed_data_manager.rs:95]: I was not aware that this was even possible, it has been fixed. I have also changed it for all the other functions and structs in the storage module where possible.

[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/compressed_data_manager.rs:95]: Fixed and I agree that it is not ideal to save it first in all cases but I think making this more efficient should be done together with making uncompressed data query-able.

File: server/src/storage/compressed_data_manager.rs [hunk 5]
```diff
@@ -81,34 +85,104 @@ impl CompressedDataManager {
...
-            let time_series_size = time_series.size_in_bytes.clone();
+    /// Save the compressed data corresponding to `key` to disk. The size of the saved compressed
+    /// data is added back to the remaining reserved memory.
+    fn save_compressed_data(&mut self, key: &u64) {
+        info!("Saving compressed time series with key '{}' to disk.", key);
 
-            let folder_path = self.data_folder_path.join(key.to_string());
-            time_series.save_to_apache_parquet(folder_path.as_path());
+        let mut time_series = self.compressed_data.remove(&key).unwrap();
+        let time_series_size = time_series.size_in_bytes.clone();
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/compressed_data_manager.rs:149]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 6]
```diff
@@ -81,34 +85,104 @@ impl CompressedDataManager {
...
+            time_series_size, self.compressed_remaining_memory_in_bytes
+        );
     }
 }
 
+/// Return [`true`] if `file_path` has a file name that is within the time range given by
+/// `start_time` and `end_time`, otherwise [`false`]. Assumes `file_path` has a file stem with
+/// the format: `start_timestamp-end_timestamp`, where both `start_timestamp` and `end_timestamp`
+/// are of the same unit as `start_time` and `end_time`.
+pub fn is_compressed_file_within_time_range(
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/compressed_data_manager.rs:167]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 7]
```diff
@@ -81,34 +85,104 @@ impl CompressedDataManager {
...
+    file_path: &PathBuf,
+    start_time: Timestamp,
+    end_time: Timestamp
+) -> bool {
+    let file_name = file_path.file_stem().unwrap().to_str().unwrap();
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/compressed_data_manager.rs:172]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 8]
```diff
@@ -222,4 +297,88 @@ mod tests {
         let data_folder_path = temp_dir.path().to_path_buf();
         (temp_dir, CompressedDataManager::new(data_folder_path))
     }
+
+    // Tests for get_saved_compressed_files().
+    #[test]
+    fn test_can_get_saved_compressed_files() {
+        let segment = test_util::get_compressed_segment_record_batch();
+        let (_temp_dir, mut data_manager) = create_compressed_data_manager();
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/compressed_data_manager.rs:305]: I prefer giving them a name for better documentation. It also makes it easier to change the code if the unused variables need to be used in the future.

File: server/src/storage/compressed_data_manager.rs [hunk 9]
```diff
@@ -222,4 +297,88 @@ mod tests {
...
+    }
+
+    #[test]
+    fn test_compressed_file_is_within_time_range() {
+        let file_path = PathBuf::from("test/1-20.parquet");
+        assert!(is_compressed_file_within_time_range(&file_path, 5, 15))
+    }
+
+    #[test]
+    fn test_compressed_file_is_not_within_time_range() {
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/compressed_data_manager.rs:380]: Added. Also fixed a minor naming issue with the tests.

File: server/src/storage/mod.rs [hunk 10]
```diff
@@ -189,9 +242,141 @@ impl StorageEngine {
...
     use tempfile::{tempdir, TempDir};
 
+    use crate::types::TimestampArray;
+
+    // Tests for get_compressed_files().
+    #[test]
+    fn test_can_get_compressed_files_from_keys() {
+        let (_temp_dir, mut storage_engine) = create_storage_engine();
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/mod.rs:256]: Fixed.

File: server/src/storage/mod.rs [hunk 11]
```diff
@@ -189,9 +242,141 @@ impl StorageEngine {
...
+        storage_engine.compressed_data_manager.insert_compressed_segment(2, segment);
+
+        let result = storage_engine.get_compressed_files(&[1, 2], None, None);
+
+        assert!(result.is_ok());
+        assert_eq!(result.unwrap().len(), 2);
+    }
+
+    #[test]
+    fn test_cannot_get_compressed_files_from_non_existent_key() {
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/mod.rs:270]: Fixed.

File: server/src/storage/time_series.rs (hunk 1 above)
Comment (CGodiksen) [reply server/src/storage/time_series.rs:85]: Fixed.

File: server/src/storage/mod.rs (hunk 2 above)
Comment (CGodiksen) [reply server/src/storage/mod.rs:248]: It has been removed.

File: server/src/storage/mod.rs [hunk 12]
```diff
@@ -325,10 +511,14 @@ pub mod test_util {
 
     /// Return a generated compressed segment with three model segments.
     pub fn get_compressed_segment_record_batch() -> RecordBatch {
+        let time_ms = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as i64;
```
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/mod.rs:514]: Good idea, I have added a `get_compressed_segment_record_batch_with_time(time_ms)` function which is now used wherever necessary.

File: server/src/storage/compressed_data_manager.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/compressed_data_manager.rs:304]: Ok

**Review by skejserjensen — CHANGES_REQUESTED** [review]
See my [reply](https://github.com/ModelarData/ModelarDB-RS/pull/34#discussion_r974988960) to this conversation.

File: server/src/storage/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/mod.rs]: Fixed.

**Review by skejserjensen — APPROVED** [review]

**Pull request merged**